File: GeoQuiz/app/src/main/java/com/davdo/geoquiz/Quiz.java
package com.davdo.geoquiz;


import android.os.Parcel;
import android.os.Parcelable;

public class Quiz implements Parcelable {

    private int mQuestionIndex;

    private int mScoreValue;

    private Question[] mQuestionList = {
        new Question(R.string.question_australia, true),
                new Question(R.string.question_oceans, true),
                new Question(R.string.question_mideast, false),
                new Question(R.string.question_africa, false),
                new Question(R.string.question_americas, true),
                new Question(R.string.question_asia, true)
    };

    private int[] mUserAnswers;

    public Quiz() {
        mQuestionIndex = 0;
        mScoreValue = 0;
    }

    private Quiz(Parcel in) {

        mQuestionIndex = in.readInt();
        mScoreValue = in.readInt();
        mQuestionList = (Question[]) in.readArray((Question[].class.getClassLoader()));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel out, int flags) {

        out.writeInt(mQuestionIndex);
        out.writeInt(mScoreValue);
        out.writeArray(mQuestionList);
    }


    public void startQuiz() {
        mUserAnswers = new int[mQuestionList.length];
        mQuestionIndex = 0;
        mScoreValue = 0;
    }

    public void setQuestion(int index) {

        if(index < 0) return;
        if(index > mQuestionList.length - 1) return;

        mQuestionIndex = index;
    }

    public int getScore() {
        return mScoreValue;
    }

    public int getQuestionIndex() {
        return mQuestionIndex;
    }

    public Question getCurrentQuestion() {
        return mQuestionList[mQuestionIndex];
    }

    public Question[] getQuestionList() {
        return mQuestionList;
    }

    public void progressQuestion() {
        progressQuestion(false);
    }

    public void progressQuestion(boolean backward) {

        if(backward)
            mQuestionIndex--;
        else
            mQuestionIndex++;

        if(mQuestionIndex < 0)
            mQuestionIndex = mQuestionList.length - 1;

        if(mQuestionIndex >= mQuestionList.length)
            mQuestionIndex = 0;

        setQuestion(mQuestionIndex);
    }

    public boolean isCurrentQuestionAnswered() {

        return isQuestionAnswered(mQuestionIndex);
    }

    public boolean isQuestionAnswered(int index) {

        return mUserAnswers[index] != 0;
    }

    public int getUserAnswer(int index) {

        return mUserAnswers[index];
    }

    protected boolean selectAnswer(boolean answer) {

        boolean correct = (getCurrentQuestion().getCorrectAnswer() == answer);

        if(correct) {
            mUserAnswers[mQuestionIndex] = 1;

            if(mScoreValue < 0)
                mScoreValue = 0;
            else
                mScoreValue++;
        }
        else {
            mUserAnswers[mQuestionIndex] = -1;
            if(mScoreValue <= 0)
                mScoreValue--;
            else
                mScoreValue = 0;
        }

        progressQuestion();

        return correct;
    }


    public static final Parcelable.Creator<Quiz> CREATOR = new Parcelable.Creator<Quiz>() {
        public Quiz createFromParcel(Parcel in) {
            return new Quiz(in);
        }

        public Quiz[] newArray(int size) {
            return new Quiz[size];
        }
    };
}

File: GeoQuizKotlin/app/src/main/java/com/davdo/geoquizkotlin/src/Quiz.kt
package com.davdo.geoquizkotlin.src

import android.os.Parcel
import android.os.Parcelable
import com.davdo.geoquizkotlin.R as Res


class Quiz() : Parcelable {

    var questionList: MutableList<Question> = mutableListOf(
        Question(Res.string.question_australia, true),
        Question(Res.string.question_oceans, true),
        Question(Res.string.question_mideast, false),
        Question(Res.string.question_africa, false),
        Question(Res.string.question_americas, true),
        Question(Res.string.question_asia, true)
    )
        set(value) {
            if (value.isNotEmpty()) {
                field = value
                score = 0
                questionIndex = 0
            }
        }

    var currentQuestion: Question
        get() {
            return questionList[questionIndex]
        }
        set(value) {
            val tempIndex = questionList.indexOf(value)
            if (tempIndex != -1)
                questionIndex = tempIndex
        }

    var questionIndex :Int = 0
        set(value) {
            field = when {
                (value >= 0 && value <= questionList.size - 1) -> value
                else -> field
            }
        }

    var score :Int = 0
        private set

    init {
        questionIndex = 0
        score = 0
    }

    constructor(questionList: MutableList<Question>): this() {

        this.questionList = questionList
    }

    private constructor(inData: Parcel) : this() {

        questionIndex = inData.readInt()
        score = inData.readInt()

        inData.readTypedList<Question>(questionList, Question)
    }

    override fun writeToParcel(out: Parcel, flags: Int) {

        out.writeInt(questionIndex)
        out.writeInt(score)

        //out.writeTypedArray(mQuestionList, 0)

        out.writeTypedList<Question>(questionList)
    }

    fun startQuiz() {
        questionIndex = 0
        score = 0
        questionList.shuffle()
    }

    fun resetQuiz() {
        startQuiz()
        for (q in questionList) {
            q.resetAnswer()
        }
    }


    fun progressQuestion() {
        progressQuestion(false)
    }

    fun progressQuestion(backward: Boolean) {

        if (backward) {
            if(questionIndex - 1 < 0)
                questionIndex = questionList.size - 1
            else
                questionIndex--
        }
        else {
            if(questionIndex + 1 > questionList.size - 1)
                questionIndex = 0
            else
                questionIndex++
        }
    }

    fun selectAnswer(answer: Boolean): Boolean {

        currentQuestion.setUserAnswer(answer)

        val correct = currentQuestion.isUserCorrect()

        if (currentQuestion.userCheated) {
            score -= 10
        }
        else {
            if (correct) score++
            else score--
        }

        return correct
    }


    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Quiz> {
        override fun createFromParcel(parcel: Parcel): Quiz {
            return Quiz(parcel)
        }

        override fun newArray(size: Int): Array<Quiz?> {
            return arrayOfNulls(size)
        }
    }
}